### PR TITLE
[MPS] Add MacOS 14 runtime check

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115514
* #115554
* #115513
* __->__ #115512

Prerequisite for adding more complex type support and FFT operation

Check using `conjugateWithTensor:name:` selector defined as follows
```objc
/// Returns the complex conjugate of the input tensor elements.
///
/// - Parameters:
///   - tensor: The input tensor.
///   - name: An optional string which serves as an identifier for the operation..
/// - Returns: A valid `MPSGraphTensor` object containing the elementwise result of the applied operation.
-(MPSGraphTensor *) conjugateWithTensor:(MPSGraphTensor *) tensor
                                   name:(NSString * _Nullable) name
MPS_AVAILABLE_STARTING(macos(14.0), ios(17.0), tvos(17.0))
MPS_SWIFT_NAME( conjugate(tensor:name:) );
```

- Rename `isOnMacOS13orNewer(unsigned minor)` hook to `isOnMacOSorNewer(major, minor)`
- Replace `torch._C.__mps_is_on_macos_13_or_newer` with `torch._C._mps_is_on_macos_or_newer`
- Add `torch.backends.mps.is_macos_or_newer` public API

### CODE REVIEW

File: aten/src/ATen/mps/MPSHooks.mm [hunk 1]
```diff
@@ -19,7 +19,16 @@
...
 
-bool MPSHooks::isOnMacOS13orNewer(unsigned minor) const {
+bool MPSHooks::isOnMacOSorNewer(unsigned major, unsigned minor) const {
+  if (major >= 14) {
+    if (major > 14) {
+      TORCH_WARN("Can't check whether running on ", major, ".", minor, "+ returning one for 14.0+");
+    } else if (minor > 0) {
+      TORCH_WARN("Can't check whether running on 14.", minor, "+ returning one for 14.0+");
+    }
+    return is_macos_13_or_newer(MacOSVersion::MACOS_VER_14_0_PLUS);
```
Comment (albanD) [inline aten/src/ATen/mps/MPSHooks.mm:29]: Should is_macos_13_or_newer be renamed as well?

File: torch/backends/mps/__init__.py [hunk 2]
```diff
@@ -22,10 +22,16 @@ def is_available() -> bool:
     return torch._C._mps_is_available()
 
 
+@_lru_cache
+def is_macos_or_newer(major: int = 13, minor: int = 0) -> bool:
+    r"""Return a bool indicating whether MPS is running on gven MacOS or newer."""
```
Comment (albanD) [inline torch/backends/mps/__init__.py:27]:
```suggestion
    r"""Return a bool indicating whether MPS is running on given MacOS or newer."""
```

File: torch/backends/mps/__init__.py [hunk 3]
```diff
@@ -22,10 +22,16 @@ def is_available() -> bool:
     return torch._C._mps_is_available()
 
 
+@_lru_cache
+def is_macos_or_newer(major: int = 13, minor: int = 0) -> bool:
```
Comment (albanD) [inline torch/backends/mps/__init__.py:26]: Do you really want default values here?

File: aten/src/ATen/mps/MPSHooks.mm (hunk 1 above)
Comment (malfet) [reply aten/src/ATen/mps/MPSHooks.mm:29]: This is part of the stable API, and `is_macos_13_or_newer(MACOS_VER_14_0_PLUS)` sounds like a reasonable API call, but sure, I can rename it

**Review by albanD — APPROVED** [review]
SGTM!

**Pull request closed without merging**